pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1170,6 +1179,9 @@ private CompletableFuture<Void> delete(boolean failIfHasSubscriptions,
                                 .map(PersistentSubscription::getName).toList();
                 return FutureUtil.failedFuture(
                         new TopicBusyException("Topic has subscriptions did not catch up: " + backlogSubs));
+            } else if (!closeIfClientsConnected && currentUsageCount() != 0 && !failIfHasBacklogs) {
+                return FutureUtil.failedFuture(new TopicBusyException(
+                        "Topic has " + currentUsageCount() + " connected producers/consumers"));
```

Comment:
It needs to fail as long as there are connections, so it has nothing to do with `!failIfHasBacklogs`.

Or if failIfHasBacklogs is triggered, the error message returned should not be like this

Reply:
Hi @codelipenghui 

When failIfHasBacklogs is true, the expected behavior is that: 

- failure of any producer exists 
- if any consumer exists, close connections

Actually, the code should be like this:

```java
if ( !closeIfClientsConnected && currentUsageCount() ){
   fail...
} else if ( !closeIfClientsConnected !failIfHasBacklogs && anyProducerExists() ){
   fail...
}
```

This code may not be easy to understand from the context, but the logic is correct

Reply:
There is also a concurrency problem that may lead to the incorrect deletion of existing producer topics. E.g:

| `checkGC` | `new producer registry`|
| --- | --- |
| ensure no producer exists |  |
| delete topic ( `false`, `true`, `false`) |  |
| ensure no backlog |  |
|  | new producer registry |
|  fence topic |  |
| disconnect all clients | |
| delete topic | |

The above flow shows the case:  producer exists but the topic is deleted. Would it be better to add a double-check？

Reply:
@poorbarcode We have write lock here https://github.com/apache/pulsar/pull/18283/files#diff-5edf14cc6f25857d0cfdd26b2d3b3141230ecfb0dfa95aebf7583fd76ede4c4bR1161

Reply:
@codelipenghui 

Ah, yes you are right